parsley/shared/src/main/scala/parsley/internal/deepembedding/backend/AlternativeEmbedding.scala

```diff
@@ -197,53 +221,74 @@ private [backend] object Choice {
     }
-    // TODO: Refactor
-    @tailrec private def foldTablified(tablified: List[(StrictParsley[_], (Char, Iterable[ExpectItem], Int, Boolean))], // scalastyle:ignore parameter.number
-                                       labelGen: CodeGenState,
-                                       roots: mutable.Map[Char, mutable.ListBuffer[StrictParsley[_]]],
-                                       backtracking: mutable.Map[Char, Boolean],
-                                       leads: mutable.ListBuffer[Char],
-                                       labels: mutable.ListBuffer[Int],
-                                       size: Int,
-                                       expecteds: List[ExpectItem],
-                                       // build in reverse!
-                                       expectedss: List[Iterable[ExpectItem]]):
-        (List[List[StrictParsley[_]]], List[Char], List[Int], Int, Iterable[ExpectItem], List[(Iterable[ExpectItem], Boolean)]) = tablified match {
-        case (root, (c, expected, _size, backtracks))::tablified_ =>
-            if (roots.contains(c)) {
-                roots(c) += root
-                backtracking(c) = backtracking(c) && backtracks
-                foldTablified(tablified_, labelGen, roots, backtracking, leads, labels,
-                              Math.max(size, _size), expected ++: expecteds, expectedss)
+
+    @tailrec private def foldTablableChars(tablified: List[TablableChar],
+                                           labelGen: CodeGenState,
+                                           roots: mutable.Map[Char, (Int, mutable.ListBuffer[StrictParsley[_]])],
+                                           map: mutable.Map[Char, (Int, Iterable[ExpectItem], Boolean)],
+                                           leads: List[Char],
+                                           size: Int,
+                                           expecteds: List[ExpectItem]):
+        (List[(Int, List[StrictParsley[_]])], mutable.Map[Char, (Int, Iterable[ExpectItem], Boolean)], Int, List[ExpectItem]) // Roots, map, size, expecteds
+            = tablified match {
+                case TablableChar(root, TablableCharDesc(c, expected, _size, backtracks)) :: tablified_ =>
+                    if (roots.contains(c)) {
+                        roots(c)._2 += root
+                        map(c) = (map(c)._1, map(c)._2, map(c)._3 && backtracks)
+                        foldTablableChars(tablified_, labelGen, roots, map, leads,
+                                                 Math.max(size, _size), expecteds ++ expected)
+                    } else {
+                        val label = labelGen.freshLabel()
+                        roots(c) = (label, mutable.ListBuffer(root))
+                        map(c) = (label, expecteds, backtracks)
+                        foldTablableChars(tablified_, labelGen, roots, map, leads :+ c,
+                                                 Math.max(size, _size), expecteds ++ expected)
```

Comment:
`size.max(_size)` I think works across 2.12 and 2.13+

Reply:
Also, this `++` is suspicious even in my own code, it's going to be a O(n^2) thing... I think actually that `++=` on `ListBuffer` will actually do this better, so it should in principle only be O(n).

Reply:
also you appear to be adding the expecteds in the reverse order to me?